msal/application.py

```diff
@@ -918,11 +918,15 @@ def _acquire_token_silent_by_finding_specific_refresh_token(
         client = self._build_client(self.client_credential, authority)
 
         response = None  # A distinguishable value to mean cache is empty
-        for entry in matches:
+        for entry in sorted(  # Since unfit RTs would not be aggressively removed,
+                              # we start from newer RTs which are more likely fit.
+                matches, key=lambda e: int(e.get("cached_at", "0")), reverse=True):
```

Comment:
Maybe we can also put tenant ID in the RT entry and use the "best match" approach.

(Actually the tenant ID and scopes are already in the key of the RT)

```
{
    "92bebeb6-c875-4f40-b2d2-f6324929f04a.54826b22-38d6-4fb2-bad9-b7b93a3e9c5a-↵
    login.microsoftonline.com-↵
    refreshtoken-↵
    04b07795-8ddb-461a-bbee-02f9e1bf7b46-↵
    -↵
    https://management.azure.com/user_impersonation https://management.azure.com/.default": {
        "credential_type": "RefreshToken",
        "secret": "...",
        "home_account_id": "92bebeb6-c875-4f40-b2d2-f6324929f04a.54826b22-38d6-4fb2-bad9-b7b93a3e9c5a",
        "environment": "login.microsoftonline.com",
        "client_id": "04b07795-8ddb-461a-bbee-02f9e1bf7b46",
        "target": "https://management.azure.com/user_impersonation https://management.azure.com/.default",        
        "family_id": "1",
        "tenant_id": "54826b22-38d6-4fb2-bad9-b7b93a3e9c5a"
    }
}
```

Reply:
The motivation of "precise match" is understandable, however the token cache schema was not capable to facilitate an `O(1)` key-value search (due to its scopes were not sorted, documented in this [internal schema doc](https://identitydivision.visualstudio.com/devex/_git/AuthLibrariesApiReview?path=%2FSSO%2FSchema.md&_a=preview&anchor=key-components)).

Hopefully, here we do not have to rely on "precise match" here. Different requests would likely converge to use the newer RT, which will naturally accumulate more and more claims.

This conversation does remind me another thing. Ideally, this PR should implement `last_modification_time` rather than `cached_at`.